clash-ghc/src-ghc/Clash/GHC/PartialEval.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Clash.GHC.PartialEval
+  ( ghcEvaluator
+  ) where
+
+import Control.Monad.State.Strict (State)
+import qualified Control.Monad.State.Strict as State
+import GHC.Integer.GMP.Internals (Integer(..))
+
+import Clash.Core.DataCon
+import Clash.Core.Evaluator.Models
+import Clash.Core.Evaluator.Semantics
+import Clash.Core.Literal
+import Clash.Core.Term
+import Clash.Core.Type
+
+ghcEvaluator :: Evaluator
+ghcEvaluator = Evaluator ghcEvaluate ghcQuote
+
+ghcEvaluate :: Term -> State Env Value
+ghcEvaluate term =
+  case term of
+    Var v -> evaluateVarWith ghcEvaluate v
+    Literal l -> evaluateLiteral l
+    Data dc -> evaluateDataWith ghcEvaluate dc
+    Prim p -> evaluatePrimWith ghcEvaluate ghcEvaluatePrim p
+    Lam x e -> evaluateLam x e
+    TyLam x e -> evaluateTyLam x e
+    App x y -> evaluateAppWith ghcEvaluate ghcEvaluatePrim ghcApply x (Left y)
+    TyApp x ty -> evaluateAppWith ghcEvaluate ghcEvaluatePrim ghcApply x (Right ty)
+    Letrec bs x -> evaluateLetrecWith ghcEvaluate bs x
+    Case x ty xs -> evaluateCaseWith ghcEvaluate ghcIsLiteral ghcIsData x ty xs
+    Cast x a b -> evaluateCastWith ghcEvaluate x a b
+    Tick ti x -> evaluateTickWith ghcEvaluate ti x
+
+-- TODO Implement evaluation for primitives and call here.
+--
+ghcEvaluatePrim :: PrimInfo -> [Either Value Type] -> State Env Value
+ghcEvaluatePrim p args = pure (VNeu $ NePrim p args)
+
+ghcIsLiteral :: Literal -> Pat -> Bool
```

Comment:
Perhaps rename this to `ghcMatchLiteral`? with some comments why it's written the way it is.